src/dfx/src/config/cache.rs

```diff
@@ -121,160 +117,98 @@ pub fn is_version_installed(v: &str) -> Result<bool, CacheError> {
-            None
-        };
-
-        let rand_string: String = thread_rng()
-            .sample_iter(&Alphanumeric)
-            .take(12)
-            .map(|byte| byte as char)
-            .collect();
-        let temp_p = get_bin_cache(&format!("_{}_{}", v, rand_string))
-            .context("Failed to get temporary bin cache path.")?;
-        std::fs::create_dir(&temp_p).with_context(|| {
-            format!(
-                "Failed to create temporary bin cache dir {}.",
-                temp_p.to_string_lossy()
-            )
-        })?;
-
-        let mut binary_cache_assets =
-            util::assets::binary_cache().context("Failed to get asset binary cache.")?;
-        // Write binaries and set them to be executable.
-        for file in binary_cache_assets
-            .entries()
-            .context("Failed to get binary cache archive entires.")?
-        {
-            let mut file = file.context("Failed to get binary cache archive entry.")?;
+    let parsed_version = Version::parse(v).map_err(CacheError::MalformedSemverVersion)?;
+    if parsed_version != *dfx_version() {
+        return Err(CacheError::UnknownVersion(v.to_owned()));
+    };
+    // Dismiss as fast as possible. We use the current_exe variable after an
+    // expensive step, and if this fails we can't continue anyway.
+    let current_exe = std::env::current_exe()?;
+
+    let b: Option<ProgressBar> = if atty::is(atty::Stream::Stderr) {
+        let b = ProgressBar::new_spinner();
+        b.set_draw_target(ProgressDrawTarget::stderr());
+        b.set_message(format!("Installing version {} of dfx...", v));
+        b.enable_steady_tick(80);
+        Some(b)
+    } else {
+        None
+    };
 
-            if file.header().entry_type().is_dir() {
-                continue;
-            }
-            file.unpack_in(temp_p.as_path())
-                .context("Failed to unpack archive asset.")?;
-            // On *nix we need to set the execute permission as the tgz doesn't include it
-            #[cfg(unix)]
-            {
-                let full_path = temp_p.join(file.path().context("Failed to get file path.")?);
-                let mut perms = std::fs::metadata(full_path.as_path())
-                    .with_context(|| {
-                        format!(
-                            "Failed to get file metadata for {}.",
-                            full_path.to_string_lossy()
-                        )
-                    })?
-                    .permissions();
-                perms.set_mode(EXEC_READ_USER_ONLY_PERMISSION);
-                std::fs::set_permissions(full_path.as_path(), perms).with_context(|| {
-                    format!(
-                        "Failed to set file permissions for {}.",
-                        full_path.to_string_lossy()
-                    )
-                })?;
-            }
-        }
+    let bin_cache_path = get_bin_cache_root()?;
+    let tempdir = tempfile::tempdir_in(bin_cache_path)?;
```

Comment:
I'd suggest to make this change as a separate PR (but really, not at all: see below).  It isn't a refactor, in the sense that changing DfxResult<T> to Result<T, Something> is a refactor, or moving a method or type from one crate to another is a refactor.  It's changing functionality.

Submitting as a separate PR will facilitate code reviews by making it easier to focus on:
- the change in functionality (tempdir_in vs thread_rng)
- the actual refactor, expected to be simpler in scope

Also, please see `list_versions()`, specifically this piece, which will not work with this change as written:
```
            if version.starts_with('_') {
                // temp directory for version being installed
                continue;
            }
```

With this in mind, I would further suggest that now is not the time for this kind of functionality change.  The thread_rng approach doesn't introduce any new error types, so it looks like it does not by itself complicate the required changes to remove dependencies on DfxResult.

Reply:
wow. first of all, thank you for catching this, I would've missed the issue with `version.starts_with('_')`. 

when I was pushing these changes my gut was telling me that including `tempfile` changes wasn't a good idea, but old habits kicked in "it's an improvement, let's smuggle it, and everything will be great". never again. and thanks for the detailed explanation!